File: types/type_ref.go
package types

import (
	"github.com/attic-labs/noms/d"
	"github.com/attic-labs/noms/ref"
)

// TypeRef structs define and describe Noms types, both custom and built-in.
// name is required for StructKind and EnumKind types, and may be allowed for others if we do type aliases. Named types are 'exported' in that they can be addressed from other type packages.
// Desc provided more details of the type. It may contain only a types.NomsKind, in the case of primitives, or it may contain additional information -- e.g. element TypeRefs for compound type specializations, field descriptions for structs, etc. Either way, checking Kind() allows code to understand how to interpret the rest of the data.
// If Kind() refers to a primitive, then Desc is empty.
// If Kind() refers to List, Map, Set or Ref, then Desc is a list of TypeRefs describing the element type(s).
// If Kind() refers to Struct, then Desc is {[name, type, ...], [name, type, ...]}.
// If Kind() refers to Enum, then Desc is a List(String) describing the enumerated values.
// If Kind() refers to an UnresolvedKind, then Desc contains a PackageRef, which is the Ref of the package where the type definition is defined. The ordinal, if not -1, is the index into the Types list of the package. If the Name is set then the ordinal needs to be found.
type TypeRef struct {
	name name
	Desc TypeDesc

	ref *ref.Ref
}

type name struct {
	namespace, name string
}

func (n name) compose() (out string) {
	d.Chk.True(n.namespace == "" || (n.namespace != "" && n.name != ""), "If a TypeRef's namespace is set, so must name be.")
	if n.namespace != "" {
		out = n.namespace + "."
	}
	if n.name != "" {
		out += n.name
	}
	return
}

// IsUnresolved returns true if t doesn't contain description information. The caller should look the type up by Ordinal in the Types of the appropriate Package.
func (t TypeRef) IsUnresolved() bool {
	_, ok := t.Desc.(UnresolvedDesc)
	return ok
}

func (t TypeRef) HasPackageRef() bool {
	return t.IsUnresolved() && !t.PackageRef().IsEmpty()
}

// Describe() methods generate text that should parse into the struct being described.
// TODO: Figure out a way that they can exist only in the test file.
func (t TypeRef) Describe() (out string) {
	if t.name != (name{}) {
		out += t.name.compose() + "\n"
	}
	if !t.IsUnresolved() {
		out += t.Desc.Describe() + "\n"
	}
	return
}

func (t TypeRef) Kind() NomsKind {
	return t.Desc.Kind()
}

func (t TypeRef) PackageRef() ref.Ref {
	desc, ok := t.Desc.(UnresolvedDesc)
	d.Chk.True(ok, "PackageRef only works on unresolved type refs")
	return desc.pkgRef
}

func (t TypeRef) Ordinal() int16 {
	d.Chk.True(t.HasOrdinal(), "Ordinal has not been set")
	return t.Desc.(UnresolvedDesc).ordinal
}

func (t TypeRef) HasOrdinal() bool {
	return t.IsUnresolved() && t.Desc.(UnresolvedDesc).ordinal >= 0
}

func (t TypeRef) Name() string {
	return t.name.name
}

func (t TypeRef) NamespacedName() string {
	return t.name.compose()
}

func (t TypeRef) Namespace() string {
	return t.name.namespace
}

func (t TypeRef) Ref() ref.Ref {
	return EnsureRef(t.ref, t)
}

func (t TypeRef) Equals(other Value) (res bool) {
	if other, ok := other.(TypeRef); ok {
		return t.Ref() == other.Ref()
	}
	return false
}

func (t TypeRef) Chunks() (out []Future) {
	v := t.Desc.ToValue()
	if v != nil {
		out = append(out, v.Chunks()...)
	}
	return
}

var typeRefForTypeRef = MakePrimitiveTypeRef(TypeRefKind)

func (t TypeRef) TypeRef() TypeRef {
	return typeRefForTypeRef
}

func MakePrimitiveTypeRef(k NomsKind) TypeRef {
	return buildType("", PrimitiveDesc(k))
}

func MakePrimitiveTypeRefByString(p string) TypeRef {
	return buildType("", primitiveToDesc(p))
}

func MakeCompoundTypeRef(kind NomsKind, elemTypes ...TypeRef) TypeRef {
	if len(elemTypes) == 1 {
		d.Chk.NotEqual(MapKind, kind, "MapKind requires 2 element types.")
	} else {
		d.Chk.Equal(MapKind, kind)
		d.Chk.Len(elemTypes, 2, "MapKind requires 2 element types.")
	}
	return buildType("", CompoundDesc{kind, elemTypes})
}

func MakeEnumTypeRef(name string, ids ...string) TypeRef {
	return buildType(name, EnumDesc{ids})
}

func MakeStructTypeRef(name string, fields []Field, choices Choices) TypeRef {
	return buildType(name, StructDesc{fields, choices})
}

func MakeTypeRef(pkgRef ref.Ref, ordinal int16) TypeRef {
	d.Chk.True(ordinal >= 0)
	return TypeRef{Desc: UnresolvedDesc{pkgRef, ordinal}, ref: &ref.Ref{}}
}

func MakeUnresolvedTypeRef(namespace, n string) TypeRef {
	return TypeRef{name: name{namespace, n}, Desc: UnresolvedDesc{ordinal: -1}, ref: &ref.Ref{}}
}

func buildType(n string, desc TypeDesc) TypeRef {
	if IsPrimitiveKind(desc.Kind()) {
		return TypeRef{name: name{name: n}, Desc: desc, ref: &ref.Ref{}}
	}
	switch desc.Kind() {
	case ListKind, RefKind, SetKind, MapKind, EnumKind, StructKind, UnresolvedKind:
		return TypeRef{name: name{name: n}, Desc: desc, ref: &ref.Ref{}}
	default:
		d.Exp.Fail("Unrecognized Kind:", "%v", desc.Kind())
		panic("unreachable")
	}
}

File: js2/src/type_ref.js
/* @flow */

'use strict';

import {isPrimitiveKind, Kind} from './noms_kind.js';
import Ref from './ref.js';
import type {NomsKind} from './noms_kind.js';

type TypeDesc = {
  kind: NomsKind;
  equals: (other: TypeDesc) => boolean;
};

class PrimitiveDesc {
  kind: NomsKind;

  constructor(kind: NomsKind) {
    this.kind = kind;
  }

  equals(other: TypeDesc): boolean {
    return other instanceof PrimitiveDesc && other.kind === this.kind;
  }
}

class UnresolvedDesc {
  _pkgRef: Ref;
  _ordinal: number;

  constructor(pkgRef: Ref, ordinal: number) {
    this._pkgRef = pkgRef;
    this._ordinal = ordinal;
  }

  get kind(): NomsKind {
    return Kind.Unresolved;
  }

  equals(other: TypeDesc): boolean {
    return other instanceof UnresolvedDesc && other._pkgRef.equals(this._pkgRef)
        && other._ordinal === this._ordinal;
  }
}

class CompoundDesc {
  kind: NomsKind;
  elemTypes: Array<TypeRef>;

  constructor(kind: NomsKind, elemTypes: Array<TypeRef>) {
    this.kind = kind;
    this.elemTypes = elemTypes;
  }


  equals(other: TypeDesc): boolean {
    if (other instanceof CompoundDesc) {
      if (this.kind !== other.kind || this.elemTypes.length !== other.elemTypes.length) {
        return false;
      }

      for (let i = 0; i < this.elemTypes.length; i++) {
        if (!this.elemTypes[i].equals(other.elemTypes[i])) {
          return false;
        }
      }

      return true;
    }

    return false;
  }
}

class StructDesc {
  fields: Array<Field>;
  union: Array<Field>;

  constructor(fields: Array<Field>, union: Array<Field>) {
    this.fields = fields;
    this.union = union;
  }

  get kind(): NomsKind {
    return Kind.Struct;
  }

  equals(other: TypeDesc): boolean {
    if (other instanceof StructDesc) {
      if (this.fields.length !== other.fields.length || this.union.length !== other.union.length) {
        return false;
      }

      for (let i = 0; i < this.fields.length; i++) {
        if (!this.fields[i].equals(other.fields[i])) {
          return false;
        }
      }

      for (let i = 0; i < this.union.length; i++) {
        if (!this.union[i].equals(other.union[i])) {
          return false;
        }
      }

      return true;
    }

    return false;
  }
}

class Field {
  name: string;
  t: TypeRef;
  optional: boolean;

  constructor(name: string, t: TypeRef, optional: boolean) {
    this.name = name;
    this.t = t;
    this.optional = optional;
  }

  equals(other: Field): boolean {
    return this.name === other.name && this.t.equals(other.t) && this.optional === other.optional;
  }
}


class TypeRef {
  _namespace: string;
  _name: string;
  _desc: TypeDesc;
  _ref: Ref;

  constructor(name: string = '', namespace: string = '', desc: TypeDesc, ref: Ref = new Ref()) {
    this._name = name;
    this._namespace = namespace;
    this._desc = desc;
    this._ref = ref;
  }

  equals(other: TypeRef): boolean {
    // TODO: Go code uses Ref() equality.
    return this.namespacedName === other.namespacedName && this._desc.equals(other._desc);
  }

  get kind(): NomsKind {
    return this._desc.kind;
  }

  get desc(): TypeDesc {
    return this._desc;
  }

  get unresolved(): boolean {
    return this._desc instanceof UnresolvedDesc;
  }

  get hasPackageRef(): boolean {
    return this.unresolved && !this.packageRef.isEmpty();
  }

  get packageRef(): Ref {
    if (this._desc instanceof UnresolvedDesc) {
      return this._desc._pkgRef;
    }

    throw new Error('PackageRef only works on unresolved type refs');
  }

  get ordinal(): number {
    if (this._desc instanceof UnresolvedDesc) {
      return this._desc._ordinal;
    }

    throw new Error('Ordinal has not been set');
  }

  get name(): string {
    return this._name;
  }

  get namespace(): string {
    return this._namespace;
  }

  get namespacedName(): string {
    let out = '';

    if (this._namespace !== '') {
      out = this._namespace + '.';
    }
    if (this._name !== '') {
      out += this._name;
    }

    return out;
  }

  get elemTypes(): Array<TypeRef> {
    if (this._desc instanceof CompoundDesc) {
      return this._desc.elemTypes;
    }

    throw new Error('Only CompoundDesc have elemTypes');
  }
}

function buildType(n: string, desc: TypeDesc): TypeRef {
  if (isPrimitiveKind(desc.kind)) {
    return new TypeRef(n, '', desc);
  }

  switch (desc.kind) {
    case Kind.List:
    case Kind.Ref:
    case Kind.Set:
    case Kind.Map:
    case Kind.Enum:
    case Kind.Struct:
    case Kind.Unresolved:
      return new TypeRef(n, '', desc);

    default:
      throw new Error('Unrecognized Kind: ' + desc.kind);
  }
}

function makePrimitiveTypeRef(k: NomsKind): TypeRef {
  return buildType('', new PrimitiveDesc(k));
}

function makeCompoundTypeRef(k: NomsKind, ...elemTypes: Array<TypeRef>): TypeRef {
  if (elemTypes.length === 1) {
    if (k === Kind.Map) {
      throw new Error('Map requires 2 element types');
    }
  } else {
    if (k !== Kind.Map) {
      throw new Error('Only Map can have multiple element types');
    }
    if (elemTypes.length !== 2) {
      throw new Error('Map requires 2 element types');
    }
  }

  return buildType('', new CompoundDesc(k, elemTypes));
}

function makeStructTypeRef(name: string, fields: Array<Field>, choices: Array<Field>): TypeRef {
  return buildType(name, new StructDesc(fields, choices));
}

function makeTypeRef(pkgRef: Ref, ordinal: number): TypeRef {
  return new TypeRef('', '', new UnresolvedDesc(pkgRef, ordinal));
}

export {CompoundDesc, Field, makeCompoundTypeRef, makePrimitiveTypeRef, makeStructTypeRef, makeTypeRef, StructDesc, TypeRef};
